### PR TITLE
7.x 1889

### PR DESCRIPTION
**Pull request by MorganDawe**
**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-1889)

# What does this Pull Request do?
When preforming a facet search on the Islandora SOLR facet pages page (ex: {site}/browse/{title}) or with the advanced search block, special characters (such as '[' or ']') will throw an 'Error searching Solr index "400" Status: Bad Request'.

# What's new?
Implements the [islandora_solr_facet_escape()|https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L165] utility function on the search term prior to query execution.

# How should this be tested?
- Navigate to a configured SOLR facet page (ex: {site}/browse/title) with the facet page search enabled.
- Preform a search query with an invalid character (such as 'Lobster[' or 'lobster{{}'). Results pages should not throw a 400 error.

# Steps to reproduce
- Navigate to a configured SOLR facet page (ex: {site}/browse/title) with the facet page search enabled, or the advanced search block
- Preform a search query with an invalid character (such as 'Lobster[' or 'lobster{{}'). Results pages will throw an error.

# Interested parties
@Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by MorganDawe** [comment]
Thanks @rosiel 👍

**Comment by DiegoPino** [comment]
@MorganDawe hi, wondering if it would not be a better idea to unify (make single function) the Solr reserved characters escaping functionality. We have https://github.com/Islandora/islandora_solr_search/blob/fe8fd3f6d4675a73c7a982fff7006c7c740ba3d2/includes/utilities.inc#L165 in the main Solr search module. So, I would say, for the sake of making both modules evaluate special chars the same way and also allowing simpler debugging, should we not better apply escaping functionality(or make Solr search better using your code) on the main Solr search module and calling it here? @Islandora/7-x-1-x-committers any thoughts on this?

**Comment by MorganDawe** [comment]
Instead of posting it here, linking to a comment which pertains to this in the islandora_solr_search module pull [here](https://github.com/Islandora/islandora_solr_search/pull/301#issuecomment-271664212)

**Comment by MorganDawe** [comment]
Refactored functionality to utilize the new regular expression pattern in the newly updated [islandora_solr_search pull](https://github.com/Islandora/islandora_solr_search/pull/301).

**Comment by DiegoPino** [comment]
@MorganDawe cool. Was waiting for this pull. Will test both during the morning (just quickly) and proceed then to merge. Thanks

**Comment by MorganDawe** [comment]
@DiegoPino If you have a free moment, could you take a quick peek at this pull?

**Comment by DiegoPino** [comment]
@MorganDawe on my list. Again, got ninjaed with new commits inside the same pull when I was already testing and, well, life went its normal route.. Will have this during the weekend.

**Comment by MorganDawe** [comment]
Pinging for review.
@DiegoPino @Islandora/7-x-1-x-committers

**Review by DiegoPino — COMMENTED** [review]
@MorganDawe I'm spinning a vagrant machine now to test this. Any edge cases I should look for as test candidates? 

Also, quick question before going further: does this form handler [here](https://github.com/Islandora/islandora_solr_facet_pages/pull/32/files#diff-9c0b24b98a747f08cc3a1de6085c065bR475) stays the same, without any modifications inside - means no escaping or further changes on the search_term values?

Just to be sure (just a question, can't criticize without having tested it), but on the similar pull on Solr search I see some escaping. https://github.com/Islandora/islandora_solr_search/pull/301/files#diff-3c926c7f5c8cd595044032e599489b9fR589 on that spot.

**Comment by DiegoPino** [comment]
@MorganDawe was waiting for your feedback on the "escaping on the search field" issue but went to testing anyway, and now after testing I'm conflicted with how the parent pull (advanced search one) is doing this instead of what is happening only in this one 😟  Let me elaborate this:

Remembering https://github.com/Islandora/islandora_solr_search/pull/301#issuecomment-277081222 
On that pull request and in that comment, Letter C) i was asking for some feedback on expected behaviour on escaping values in the search fields. The following relates to that message and to my own message here yesterday.

So i did a simple test last night:

1. Create an object with title `\\Object`
2. Enabled escaping on solr facet pages and  advanced search, configured for both `fgs_label_s` as my only field.
3. I added all search fields, simple, advanced and facet side by side
4. I searched for `\\Object` in all fields/modules

Escaping is inconsistent in UI.
 * Simple search gives correct results, breadcrumb shows `\Object`, if I search for `\\\\Object` it shows `\\Object`. Search field does not maintain value.
 * Advanced Search, gives results, search field gets its value changed (see #301) and one backslash removed. That is inconsistent because if you happen to press search again, it removes the second backslash... so it is not respecting the users choice on what he wants to search for.
  * Facet pages does not show `\` as part of the alphabet list (so it is doing escaping on that list...which is deceiving) and the when searching for`\\Object` it gives me almost correct result.. (with escaping... wich means `\\Object` appears under letter `o`. But the search field itself stays at `\\Object`, which seems correct! But inconsistent with all other behaviors.

Will ping @adam-vessey @whikloj @willtp87, for their advice on this:
Also @jonathangreen @mjordan (updated, will include people that were not previously on this pulls)

Islandora UI should be consistent. In all our search interfaces the behavior should be the same. Breadcrumbs are to preserve the searchers "choice" visually (hint)... search field values, if maintained, should match what the user choose to put there, and, the results should match the user expectation. I feel the escaping strategies are there to ensure that the use can indeed search for special characters, but that must be fed back by consistent UI/UX

Thanks,😀i know it is a complicated thing test and takes time and sorry for the delay, but search interface is probably the cornerstone of a repo and requires being picky. So if needed we could have a talk during committers call or solve it here.

**Comment by MorganDawe** [comment]
@DiegoPino Thank you for testing, i was preparing a response when your latest comment popped up. I am going to review your findings before i try to elaborate more. And i agree, this type of work is the cornerstone of content discovery in Islandora, so best to get it right :)

**Comment by DiegoPino** [comment]
@MorganDawe hi, any updates on the issues? Would love to have fixes / the whole set tested before code freeze. If not, we will have to find a strategy to roll some stuff back to avoid the issue of UI inconsistency. Ping me if you need anything. Thanks!

**Comment by MorganDawe** [comment]
@DiegoPino I am sorry to keep you waiting, but i have been running some tests myself and preparing a pull request. To address the advanced search UI/UX issue, i will need to issue another pull request to the islandora_solr_search module. My goal is to have a pull request for each repo updated/submitted by the end of the day. Thanks again.

**Comment by DiegoPino** [comment]
@MorganDawe no worries. Happy to know this is advancing and thanks for your effort on this. Count on my help for any testing and of course merging. great work!

**Comment by MorganDawe** [comment]
@DiegoPino Preparing pull requests to address these issues, testing has prompted me to create [This pull](https://github.com/Islandora/islandora_solr_search/pull/311) to the islandora solr search module. I am in the process of updating this pull request to match.

**Comment by DiegoPino** [comment]
@MorganDawe ping me when you have all pieces in place and ready for testing. Thanks!

**Comment by MorganDawe** [comment]
@DiegoPino ,
I have been able to confirm your findings, and i believe between the updates to this pull, and the newly added [pull 311](https://github.com/Islandora/islandora_solr_search/pull/311) to islandora solr search, we should be able to address these issues:

* Simple Search...    - Modified the stripslashes() implementations so what is displayed to the user is the original value.
* Advanced Search ... - Given the same treatment as Simple Search
* Facet Pages ...     - With the latest work, Created an object with the label `'\\Object'`. Created a facet browse page using the `'fgs_label_s'` field. Searched for term `'\\Object'`. Result set shows (under 'All'), the first and only entry (in my case) is `'\\Object'` (Not escaped, for the sake of UI/UX). No entry is shown under 'O', because the term is not escaped when presented to the user.

The work in this pull will require [this pull 311](https://github.com/Islandora/islandora_solr_search/pull/311) to display properly as well.

This pull primarily updated the const name, so that it matches properly now with what is in the islandora_solr_search module.

This should be ready for another test pass @DiegoPino when you get a chance. We are still reviewing internally as well, so something may still pop up.

I have some time set aside tomorrow to address issues that may arise.

NOTE: For the sake of sanity, i recommend ensuring the value used locally as the escaping regex pattern is proper. If this module (and islandora solr search) are already present on a test box, the previous value for the escaping regex may not be proper. Just a heads up to double check.

**Comment by DiegoPino** [comment]
@MorganDawe cool. Will take me probably a while (tonight) to test, but if all goes right we will have this merged by tomorrow. Hope you have some time tomorrow in case something goes wrong. I really appreciate the dedication to this issue. thanks a lot. Will keep you informed!

**Comment by DiegoPino** [comment]
@MorganDawe I did all type of tests: will keep all the conversation about all the involved pieces here if possible. There are some issues 😁  so in the meantime, please do the following:

Spin new vagrant
Apply this and solr facet pages  to you vagrant machine
Create 3 new objects with titles
//Morgan
\\Morgan
and \Morgan/

Configure/enable Solr to escape regex on advanced search and facet pages
Configure Solr index to display fgs_label_s and pid in search results, add fgs_label_s to facets (i know..)
Configure Solr facet pages for  fgs_label_s

Make all visible in structure.

That for now. Will touch base in an hour

**Comment by DiegoPino** [comment]
@MorganDawe @adam-vessey and @Islandora/7-x-1-x-committers in general. Lets see if we fix this before code freeze:

Notes: consistency 👍 means people get back/hints what they searched for in links, breadcrumbs, etc



 * Simple search with forward slash: search 👍 consistency 👍 . This was not affected by the multiple pulls.
simple search forward slash


 * Advanced search with forward Slash: search ERROR 👎 consistency 👍  *partial, we don't have more info to compare against because of the error
advanced search forward slash

* Advanced search with double forward Slash: No error, but search results == 0 / should be one👎 consistency 👍
advanced search 2x forward slash
* Advanced search with backward Slash: Search results  👍  consistency 👎 see extra backward slashes added to breadcrumb and query
advanced search backward slash

**Comment by DiegoPino** [comment]
And now facetpages:
First 2 ones are wrong:

facet search forward
facet search single forward


This is right 👍  

face page search backward

Also, wonder why the landing page(click on the result of that search) strategy is different than in advanced search. It uses filters so no extra backslashes are added, which was happening in my previous comment : advanced search using fgs_label_s as search field with same query value

**Comment by willtp87** [comment]
Thanks @DiegoPino. @MorganDawe  is out today, @adam-vessey  and I are going to work on sorting this.

**Comment by willtp87** [comment]
@DiegoPino: @adam-vessey  and I are unable to reproduce the Solr error with two forward slashes. We have tried to both on Solr 4.10.3 and Solr 3.6.2.

**Comment by DiegoPino** [comment]
Only thing I did was:

- vagrant up new machine
- this commands https://gist.github.com/DiegoPino/f85f80276f8c6d88a33e4ffa6abc6351
- add a new object with label "/large//image"
- clear cache
- run update modules just in case
- enable regex escaping with the defaults for advanced.
advanced search gives this output https://gist.github.com/DiegoPino/187993458872196a28884183e85cbc41

Clueless.

**Comment by adam-vessey** [comment]
@DiegoPino: Solr 4.0+ introduced regex queries, which I suspect may be related... 4.10+ might have been made to be more graceful than the 4.2 in islandora_vagrant? Digging a bit more here.

**Comment by DiegoPino** [comment]
@adam-vessey probably. I do (as most committers I guess) tests on the vanilla islandora vagrant  machine, being all standard except for that `drush pm-update` because, well,  7.44 is just too old. Also because i really don't want to introduce any of my local fanciness into the mixture. Can further test on an 4.10 upgraded solr or even 5.x but documentation suggests we are still offering support for 4.2 
Have to dig deeper.

**Comment by willtp87** [comment]
@DiegoPino  we have been able to verify that this is a version specific issue; the exact version range is unknown.  I think we should handle it via documentation. We are testing for an alternative regex now.

**Comment by DiegoPino** [comment]
@willtp87 excellent (thanks) and bad at the same time, edge case it seems to be?. Appreciate your efforts.

**Comment by willtp87** [comment]
@DiegoPino  it seems escaping the slash in Solr 3 will cause no issue and will perhaps be good to do in Solr 4.10. So we're updating the default in the other pull.

**Comment by DiegoPino** [comment]
@willtp87 ok. Travis is still running on that, i see a new default. Do these changes solve all issues (breadcrump inconsistency included?) or are they still other to-be-made commits on their way? Thanks

**Comment by willtp87** [comment]
@DiegoPino I think it is all handled now.

**Comment by DiegoPino** [comment]
@willtp87 thanks a lot (also @adam-vessey). Will test tonight. AWS S3 downtime is affecting everything, Travis-CI and inclusive git file attachments (tried to upload a screenshot here a few times without luck) so will wait until that gets solved. Cheers

**Comment by DiegoPino** [comment]
@willtp87 @adam-vessey  if you are ok, i will merge this one in the meantime. Until we have green light on https://github.com/Islandora/islandora_solr_search/pull/312#issuecomment-283462390

**Review by DiegoPino — APPROVED** [review]
Depends on https://github.com/Islandora/islandora_solr_search/pull/312#issuecomment-283462390 to work correctly.

**Comment by adam-vessey** [comment]
@DiegoPino: https://github.com/Islandora/islandora_solr_search/pull/312 should be good to go, as of https://github.com/Islandora/islandora_solr_search/pull/312/commits/faf24e192e7275b776fa9acc292e502eab0011ca

**Comment by willtp87** [comment]
Thank you @DiegoPino for all your work on these pulls.

**Comment by DiegoPino** [comment]
@willtp87 @adam-vessey and to you guys for stepping up and solving all these issues. Was a good team effort. appreciate it.

**Pull request merged**
